### ISSUE TITLE
catch error from createAsyncThunk

### ISSUE THREAD

**Issue by affanshahid** [reduxjs/redux-toolkit#520]
So lets say I have a thunk, which I created as follows:

```ts
export const login = createAsyncThunk(
  'auth/login',
  async (credentials: Credentials): Promise<User> => {
    return authClient.login(credentials.username, credentials.password);
  }
);
```

What I usually do is in my codebase when dispatching thunks:

```ts
async function handleClick() {
  try {
    await dispatch(login(credentials))
  } catch (err) {
    showToastForError(err)
  }
}
```

However `createAsyncThunk()` does not throw the error if it occurred (instead it uses it to dispatch the rejected action). Is there some way I can achieve this? or what would be the alternative.

Also when using typescript the result of awaiting `dispatch(login(credentials))` returns an action object, but it does not have an error property which I could use to check if an error occurred. This seems just to be a typings issue since logging the actual object does show the error property exists

**Comment by markerikson** [comment]
Yep, this is covered in the docs - see https://redux-toolkit.js.org/api/createAsyncThunk#handling-thunk-results .

**Comment by mhienle** [comment]
We were facing the same issue regarding toasts in our project. We solved the issue by creating a wrapper for the async logic inside `createAsyncThunk` which in a simplified version looks as follows:

```ts
export function withToastForError<Args, Returned>(payloadCreator: (args: Args) => Promise<Returned>) {
    return async (args: Args) => {
        try {
            return await payloadCreator(args);
        } catch (err) {
            showToastForError(err);
            throw err; // throw error so createAsyncThunk will dispatch '/rejected'-action
        }
    };
}
```

In the above example, one would define the thunk as follows:

```ts
export const login = createAsyncThunk(
    'auth/login',
    withToastForError(
        async (credentials: Credentials): Promise<User> => {
            return authClient.login(credentials.username, credentials.password);
        },
    ),
);
```

and in the component

```ts
const handleClick = () => dispatch(login(credentials));
```

**Comment by RRaideRR** [comment]
```
export const login = createAsyncThunk(
    'auth/login',
    withToastForError(
        async (credentials: Credentials): Promise<User> => {
            return authClient.login(credentials.username, credentials.password);
        },
    ),
);
```

Could you explain how to access thunkAPI in this example? Sorry I can't get it to work @mhienle

**Comment by mhienle** [comment]
Accessing `thunkAPI` is one of those things I excluded in the above example (like I said, it's a simplified version 😉).

It is important to note that `withToastForError` returns the same type that it takes as parameter, i.e. a function `(args: Args) => Promise<Returned>`. It does not change the return type or return value of `payloadCreator`, it only returns the same thing as the original payloadCreator, but depending on whether that throws an error or not, it performs side effects like showing toasts.

As you see, what we do is pass the return value of `withToastForError` as the second argument to `createAsyncThunk`. So our non-simplified version of `withToastForError` does not take `(payloadCreator: (args: Args) => Promise<Returned>)` as an argument, but instead takes exactly what `createAsyncThunk` requires as second argument namely `payloadCreator: AsyncThunkPayloadCreator<Returned, ThunkArg, ThunkApiConfig>`, and this type exposes thunkAPI.

So it looks somewhat like this:
```
export function withToastForError<Args, Returned, ThunkApiConfig>(payloadCreator: AsyncThunkPayloadCreator<Returned, ThunkArg, ThunkApiConfig>) {
    return async (args, thunkAPI) => {
        try {
            return await payloadCreator(args, thunkAPI);
        } catch (err) {
            showToastForError(err);
            throw err; // throw error so createAsyncThunk will dispatch '/rejected'-action
        }
    };
}

export const login = createAsyncThunk(
    'auth/login',
    withToastForError(
        async (credentials: Credentials, thunkAPI): Promise<User> => {
            // now I should be able to use thunkAPI here
            return authClient.login(credentials.username, credentials.password);
        },
    ),
);
```

**Comment by RRaideRR** [comment]
Thanks a lot @mhienle that you took the time to answer my question in such great detail. I'm feeling much more confident right now thanks to you. Thank you :-).

**Comment by muz3** [comment]
@mhienle, you had an elegant solution. But why do you want to put your UI render logic inside createAsyncThunk. Should you just have your State to store the error. Then have your main component to check error state there?

**Comment by phryneas** [comment]
@muz3 since toasts are often timing-related and managed from start to end by third-party libraries, I think it can be totally okay to view them as a "side effect" rather than "ui-specific logic" ;)

**Comment by henrik** [comment]
Hope it's OK to ask this related question: Is there a recommended way to handle errors from inside the reducer case?

So say I have

```
.addCase(fetchThing.pending, (state, _action) => {
  state.pending = true
})
.addCase(fetchThing.fulfilled, (state, _action) => {
  functionThatUnexpectedlyThrowsAnError()
  state.pending = false
})
```

When there is an error, the state remains pending forever, with a logged error.

I believe ErrorBoundaries don't see this since it's outside the React component stack. And `fetchThing.rejected` doesn't come into it since we've already determined the outcome of `fetchThing`.

I could just use a try/catch around `functionThatUnexpectedlyThrowsAnError()`, of course, but I'd rather not sprinkle those everywhere. Would be nice if it could be centralised somehow, so any failure in any `addCase` would be handled.

**Comment by phryneas** [comment]
Tbh, you should not have any logic in there that could potentially throw an error.

This will probably result in a rejection of the promise returned by `dispatch(thunk())` but I would not 100% rely on that.

**Comment by henrik** [comment]
Thank you, @phryneas!

Saw this when I caused a null error in dev and it showed a "pending" spinner forever instead of a "something went wrong" type message. It's hard to guarantee error-free code 😁 

Good shout about `dispatch` returning a promise that is rejected. I just verified by reproducing that error in dev. Indeed, if I do `dispatch(…).catch(e => console.log("boom", e))`, it will log the error. So it seems that mechanism could be used for error handling.
